subworkflows/automate_input/bin/check_samplesheet.py

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env python
+
+# This script is based on the example at: https://raw.githubusercontent.com/nf-core/test-datasets/viralrecon/samplesheet/samplesheet_test_illumina_amplicon.csv
+
+import os
+import sys
+import errno
+import argparse
+
+
+def parse_args(args=None):
+    Description = "Reformat nf-core/readmapping samplesheet file and check its contents."
+    Epilog = "Example usage: python check_samplesheet.py <FILE_IN> <FILE_OUT>"
+
+    parser = argparse.ArgumentParser(description=Description, epilog=Epilog)
+    parser.add_argument("FILE_IN", help="Input samplesheet file.")
+    parser.add_argument("FILE_OUT", help="Output file.")
+    return parser.parse_args(args)
+
+
+def make_dir(path):
+    if len(path) > 0:
+        try:
+            os.makedirs(path)
+        except OSError as exception:
+            if exception.errno != errno.EEXIST:
+                raise exception
+
+
+def print_error(error, context="Line", context_str=""):
+    error_str = "ERROR: Please check samplesheet -> {}".format(error)
+    if context != "" and context_str != "":
+        error_str = "ERROR: Please check samplesheet -> {}\n{}: '{}'".format(
+            error, context.strip(), context_str.strip()
+        )
+    print(error_str)
+    sys.exit(1)
+
+
+def check_samplesheet(file_in, file_out):
+    """
+    This function checks that the samplesheet follows the following structure:
+
+    sample,datatype,datafile
+    sample1,hic,/path/to/file1.cram
+    sample1,illumina,/path/to/file2.cram
+    sample1,pacbio,/path/to/file3.cram
+    sample1,ont,/path/to/file4.cram
+
+    For an example see:
+    https://raw.githubusercontent.com/nf-core/test-datasets/viralrecon/samplesheet/samplesheet_test_illumina_amplicon.csv
+    """
+
+    sample_mapping_dict = {}
+    with open(file_in, "r") as fin:
+
+        ##* Check header
+        MIN_COLS = 3
+        HEADER = ["sample", "datatype", "datafile"]
+        header = [x.strip('"') for x in fin.readline().strip().split(",")]
+        if header[: len(HEADER)] != HEADER:
+            print("ERROR: Please check samplesheet header -> {} != {}".format(",".join(header), ",".join(HEADER)))
+            sys.exit(1)
+
+        ## Check sample entries
+        for line in fin:
+            lspl = [x.strip().strip('"') for x in line.strip().split(",")]
+
+            # Check valid number of columns per row
+            if len(lspl) < len(HEADER):
```

Comment:
Rather than silently discarding the extra columns
```suggestion
            if len(lspl) != len(HEADER):
```

Reply:
Having the `<` sign allows for optional columns. For example, the `library` column in `readmapping` pipeline was optional. I am not sure it would work the same way with `!=`.

Reply:
No it wouldn't, but that's the principle. If the number of columns is fixed, `check_samplesheet.py` needs to fail if the number of columns on any lines (actually including the header, cf line 61) differs. If some columns are optional, `check_samplesheet` needs to know which ones, and actually check that the number of columns is within the expected range

Reply:
Discussed this afternoon: leave `<` as it was, as the nf-core convention is to allow extra columns (that could be used in other pipelines)